### ISSUE TITLE
Generate Report dynamically - dashboard state code generator

### ISSUE THREAD

**Issue by angelikileonida** [thingsboard/thingsboard#4370]
**Component**
 * Rule Engine

We would like to generate a dashboard state report in the rule chain dynamically.  The dashboard state-input is the "entity" from the root dashboard.

We would like to have a custom action button that the user will request a report to be generated based on the entity(assset) 
and in the rule chain the specific report for a dashboard state of this entity. 

From the generate report widget seems that after user selects dashboard state for the report a dashboard state parameter value code is generated. 
![image](https://user-images.githubusercontent.com/45817004/113675349-f057b100-96c3-11eb-8f43-ddfc820fdac6.png)

Is there a way to dynamically generate the "dashboard state parameter value"  per entity (asset) in the rule chain?

from the rule chain msg generated: 
{
    "reportConfig": {
        "baseUrl": "https://www********.com",
        "useDashboardTimewindow": true,
        "timewindow": {
            "selectedTab": 1,
            "history": {
                "historyType": 0,
                "interval": 1000,
                "timewindowMs": 86400000
            }
        },
        "namePattern": "shipment-report-%d{yyyy-MM-dd_HH:mm:ss}",
        "type": "pdf",
        "timezone": "Europe/Bucharest",
        "useCurrentUserCredentials": true,
        "userId": "e1ef8d60-7ce5-11eb-a531-6b808ece1fee",
        "dashboardId": "3cb8a180-803f-11eb-a531-6b808ece1fee",
        **"state": "W3siaWQiOiJkZWZhdWx0IiwicGFyYW1zIjp7fX0seyJpZCI6IlNoaXBtZW50cyIsInBhcmFtcyI6e319LHsiaWQiOiJfc2hpcG1lbnRfZGV0YWlsc18iLCJwYXJhbXMiOnsiZW50aXR5SWQiOnsiaWQiOiJkZDI3NDgwMC04MWE4LTExZWItYTUzMS02YjgwOGVjZTFmZWUiLCJlbnRpdHlUeXBlIjoiQVNTRVQifSwiZW50aXR5TmFtZSI6IlNDLTExMDAtNzIzMDg2Mjg4NzEwMjc0MTIiLCJlbnRpdHlMYWJlbCI6IlNDLTExMDAtNzIzMDg2Mjg4NzEwMjc0MTIifX1d"**
    },
    "sendEmail": false
}

**Description**

**Environment**

 * OS:  Ubuntu
 * ThingsBoard:3.2.2PE
 * Browser: chrome, Microsoft Edge

**Comment by ashvayka** [comment]
Yes, the state is Base64 encoded JSON. For example, your state is the following:

```json
[{"id":"default","params":{}},{"id":"Shipments","params":{}},{"id":"_shipment_details_","params":{"entityId":{"id":"dd274800-81a8-11eb-a531-6b808ece1fee","entityType":"ASSET"},"entityName":"SC-1100-72308628871027412","entityLabel":"SC-1100-72308628871027412"}}]
```

I assume you can develop a custom rule node that prepares the state parameter and encodes it for you.  see development guide: https://thingsboard.io/docs/user-guide/contribution/rule-node-development/

**Comment by angelikileonida** [comment]
Hi, 

thank you for the response.  we will try that.

**Comment by angelikileonida** [comment]
Hi, 

thank you for the help. we created the state code from custom widget action and store it as an asset parameter and then trigger the rule chain for the generation of the report.
